windows/datetimepicker.cpp

```diff
@@ -131,6 +134,72 @@ static void uiDateTimePickerMinimumSize(uiWindowsControl *c, int *width, int *he
 	*height = y;
 }
 
+static BOOL onWM_NOTIFY(uiControl *c, HWND hwnd, NMHDR *nmhdr, LRESULT *lResult)
+{
+	uiDateTimePicker *d = uiDateTimePicker(c);
+
+	if (nmhdr->code != DTN_DATETIMECHANGE)
+		return FALSE;
+	(*(d->onChanged))(d, d->onChangedData);
+	*lResult = 0;
+	return TRUE;
+}
+
+static void fromSystemTime(LPSYSTEMTIME systime, struct tm *time)
+{
+	ZeroMemory(time, sizeof(struct tm));
+	time->tm_sec = systime->wSecond;
+	time->tm_min = systime->wMinute;
+	time->tm_hour = systime->wHour;
+	time->tm_mday = systime->wDay;
+	time->tm_mon = systime->wMonth - 1;
+	time->tm_year = systime->wYear - 1900;
+	time->tm_wday = systime->wDayOfWeek;
+	time->tm_isdst = -1;
+}
+
+static void toSystemTime(const struct tm *time, LPSYSTEMTIME systime)
+{
+	ZeroMemory(systime, sizeof(SYSTEMTIME));
+	systime->wYear = time->tm_year + 1900;
+	systime->wMonth = time->tm_mon + 1;
+	systime->wDayOfWeek = time->tm_wday;
+	systime->wDay = time->tm_mday;
+	systime->wHour = time->tm_hour;
+	systime->wMinute = time->tm_min;
+	systime->wSecond = time->tm_sec;
+}
+
+static void defaultOnChanged(uiDateTimePicker *d, void *data)
+{
+	// do nothing
+}
+
+void uiDateTimePickerTime(uiDateTimePicker *d, struct tm *time)
+{
+	SYSTEMTIME systime;
+
+	if (SendMessageW(d->hwnd, DTM_GETSYSTEMTIME, 0, (LPARAM) &systime) != GDT_VALID)
+		logLastError(L"error getting date and time");
+	fromSystemTime(&systime, time);
+}
+
+void uiDateTimePickerSetTime(uiDateTimePicker *d, const struct tm *time)
+{
+	SYSTEMTIME systime;
+
+	toSystemTime(time, &systime);
+	if (!SendMessageW(d->hwnd, DTM_SETSYSTEMTIME, GDT_VALID, (LPARAM) &systime))
+		logLastError(L"error setting date and time");
+	(*(d->onChanged))(d, d->onChangedData);
```

Comment:
`Set` functions should not call event handlers (yet).

Reply:
Agreed, need some sort of "properties" API long term to do this correctly. In this case though, this is simply how the control works. We call the event handler manually on Windows because its done for us on the other platforms. My example program makes straight forward use of this functionality.